src/arch_ppc.py

```diff
@@ -430,45 +474,200 @@ def parse(
+                inputs = [args[0], args[1], args[2]]
+                outputs = [MemoryAccess(args[1], args[2], size), args[1]]
+            else:
+                assert len(args) == 2 and isinstance(args[1], AsmAddressMode)
+                inputs = [args[0], args[1].rhs]
+                outputs = [make_memory_access(args[1]), args[1].rhs]
+        elif mnemonic in cls.instrs_load_update:
+            assert isinstance(args[0], Register) and size is not None
+            if mnemonic.endswith("x"):
+                assert (
+                    len(args) == 3
+                    and isinstance(args[1], Register)
+                    and isinstance(args[2], Register)
+                )
+                inputs = [MemoryAccess(args[1], args[2], size), args[1], args[2]]
+                outputs = [args[0], args[1]]
+            else:
+                assert len(args) == 2 and isinstance(args[1], AsmAddressMode)
+                inputs = [make_memory_access(args[1]), args[1].rhs]
+                outputs = [args[0], args[1].rhs]
+        elif mnemonic in ("stmw", "lmw"):
+            assert (
+                len(args) == 2
+                and isinstance(args[0], Register)
+                and isinstance(args[1], AsmAddressMode)
+                and args[0].register_name[0] == "r"
+            )
+            index = int(args[0].register_name[1:])
+            offset = args[1].lhs_as_literal()
+            while index <= 31:
+                reg = Register(f"r{index}")
+                mem = make_memory_access(
+                    AsmAddressMode(rhs=args[1].rhs, lhs=AsmLiteral(offset))
+                )
+                if mnemonic == "stmw":
+                    inputs.append(reg)
+                    outputs.append(mem)
+                else:
+                    outputs.append(reg)
+                    inputs.append(mem)
+                index += 1
+                offset += 4
+            inputs.append(args[1].rhs)
+        elif mnemonic in cls.instrs_no_dest:
+            assert not any(isinstance(a, (Register, AsmAddressMode)) for a in args)
+        elif mnemonic.rstrip(".") in cls.instrs_destination_first:
+            assert isinstance(args[0], Register)
+            outputs = [args[0]]
+            if mnemonic.startswith("l") and size is not None:
+                if mnemonic.endswith("x"):
+                    assert (
+                        len(args) == 3
+                        and isinstance(args[1], Register)
+                        and isinstance(args[2], Register)
+                    )
+                    inputs = [args[1], args[2], MemoryAccess(args[1], args[2], size)]
+                else:
+                    assert len(args) == 2 and isinstance(args[1], AsmAddressMode)
+                    inputs = [args[1].rhs, make_memory_access(args[1])]
+            elif mnemonic == "li" and isinstance(args[1], AsmAddressMode):
+                # `li $rD, sym@sda21(r2)` is equivalent to `addi $rD, $r2, sym@sda21`
```

Comment:
`li` can't read from memory, surely? looks from the implementation that we'd crash on that, too

Reply:
Ah, this was a quirk of the disassembler.
`li` is an alias for `addi`, so I thought this was a weird-but-valid form. However, `li $rD, sym@sda21(r2)` should have been disassembled as an *la* instruction, not *li*. 

[Reference, page 23](https://www.nxp.com/docs/en/application-note/AN2491.pdf)

I fixed the disassembler to emit `addi $rD, $r2, sym@sda21` instead of `li $rD, sym@sda21(r2)`, regenerated the test, and removed the above block of code. The test output doesn't change.

Although I could have changed it to `la`, and added support for `la` to PPC, but the decomp projects seem to all prefer the `addi` form instead of the `la` form